presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java

```diff
@@ -0,0 +1,231 @@
+    {
+        MemoryTableHandle memoryTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        return memoryTableHandle.toTableMetadata();
+    }
+
+    @Override
+    public List<SchemaTableName> listTables(ConnectorSession session, String schemaNameOrNull)
+    {
+        if (schemaNameOrNull != null && !schemaNameOrNull.equals(SCHEMA_NAME)) {
+            return ImmutableList.of();
+        }
+        return tables.values().stream()
+                .map(MemoryTableHandle::toSchemaTableName)
+                .collect(toList());
+    }
+
+    @Override
+    public Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        MemoryTableHandle memoryTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        return memoryTableHandle.getColumnHandles().stream()
+                .collect(toMap(MemoryColumnHandle::getName, Function.identity()));
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        MemoryColumnHandle memoryColumnHandle = checkType(columnHandle, MemoryColumnHandle.class, "columnHandle");
+        return memoryColumnHandle.toColumnMetadata();
+    }
+
+    @Override
+    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        return tables.values().stream()
+                .filter(table -> prefix.matches(table.toSchemaTableName()))
+                .collect(toMap(MemoryTableHandle::toSchemaTableName, handle -> handle.toTableMetadata().getColumns()));
+    }
+
+    @Override
+    public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        MemoryTableHandle handle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        Long tableId = tableIds.remove(handle.getTableName());
+        if (tableId != null) {
+            tables.remove(tableId);
+        }
+    }
+
+    @Override
+    public void renameTable(ConnectorSession session, ConnectorTableHandle tableHandle, SchemaTableName newTableName)
+    {
+        MemoryTableHandle oldTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        MemoryTableHandle newTableHandle = new MemoryTableHandle(
+                oldTableHandle.getConnectorId(),
+                oldTableHandle.getSchemaName(),
+                newTableName.getTableName(),
+                oldTableHandle.getTableId(),
+                oldTableHandle.getColumnHandles(),
+                oldTableHandle.getHosts(),
+                oldTableHandle.getActiveTableIds()
+        );
+        tableIds.remove(oldTableHandle.getTableName());
+        tableIds.put(newTableName.getTableName(), oldTableHandle.getTableId());
+        tables.remove(oldTableHandle.getTableId());
+        tables.put(oldTableHandle.getTableId(), newTableHandle);
```

Comment:
Thread safety for `tables` and `tableIds` are concerning.

* Sequence of `tables.remove(x); tables.put(x, ...);` can result in the table being pruned in the middle.
* `ImmutableSet.copyOf(tableIds.values)` may observe mixed state. It is not straightforward to reason about whether such observation is guaranteed to be acceptable or not.

Back to this comment about unintended deletion I left during my first-pass: https://github.com/prestodb/presto/pull/5809#discussion_r88376531

* I came up with those examples under the impression that update to `tableIds` and `tables` are carried out during `finishCreateTable` and `finishInsert`. I can see that this is not the case now. It looks like my memory didn't serve me well when I made those examples.
* The general problem still exists with the current code however. In `beginCreateTable`, increment to `nextTableId` and insertion to `tableIds` can interleave. As a result, smaller table ids are not guaranteed to be inserted into the maps prior to the larger table ids. This can lead to unintended deletion of table data (i.e. with some tweaks, the scenarios I described in the previous comment can happen because of this bullet point.)

It seems like it would just be so much easier to reason about these things if we protect all these three variables with a synchronized block. It doesn't look like concurrency performance is a concern for this class.